cmd/source_profile.go

```diff
@@ -55,27 +56,45 @@ type SourceProfileConnectionMySQL struct {
 	pwd  string // Same as MYSQLPWD environment variable
 }
 
-func NewSourceProfileConnectionMySQL(params map[string]string) SourceProfileConnectionMySQL {
-	// TODO: Move parsing of environment variables in this function.
+func NewSourceProfileConnectionMySQL(params map[string]string) (SourceProfileConnectionMySQL, error) {
 	mysql := SourceProfileConnectionMySQL{}
-	if host, ok := params["host"]; ok {
-		mysql.host = host
-	}
-	if port, ok := params["port"]; ok {
-		mysql.port = port
-	} else { // Set default port for mysql, which rarely changes.
-		mysql.port = "3306"
+	host, hostOk := params["host"]
+	user, userOk := params["user"]
+	db, dbOk := params["db_name"]
+	port, portOk := params["port"]
+	pwd, pwdOk := params["password"]
+	if !(hostOk || userOk || dbOk || portOk || pwdOk) {
+		// No connection params provided through source-profile. Fetching from env variables.
+		fmt.Printf("Connection parameters not specified in source-profile. Reading from " +
+			"environment variables MYSQLHOST, MYSQLUSER, MYSQLDATABASE, MYSQLPORT, MYSQLPWD...\n")
+		mysql.host = os.Getenv("MYSQLHOST")
+		mysql.user = os.Getenv("MYSQLUSER")
+		mysql.db = os.Getenv("MYSQLDATABASE")
+		mysql.port = os.Getenv("MYSQLPORT")
+		mysql.pwd = os.Getenv("MYSQLPWD")
+	} else if hostOk && userOk && dbOk {
+		// If atleast host, username and dbname are provided through source-profile,
+		// go ahead and use source-profile. Port and password handled later even if they are empty.
+		mysql.host, mysql.user, mysql.db, mysql.port, mysql.pwd = host, user, db, port, pwd
+	} else {
+		// Partial params provided through source-profile. Ask user to provide all.
+		return mysql, fmt.Errorf("please specify host, port, user and db_name in the source-profile")
 	}
-	if user, ok := params["user"]; ok {
-		mysql.user = user
+
+	// Throw same error if the input entered is empty.
+	if mysql.host == "" || mysql.user == "" || mysql.db == "" {
+		return mysql, fmt.Errorf("found empty string for host/user/db. please specify host, port, user and db_name in the source-profile")
```

Comment:
This is incorrect since we could reach here if let's say a user has set only some of the environment variables but not all the required ones. I suggest that we rewrite the conditional such that we make the distinction whether we are reading from source-profile vs environment variables and handle those appropriately. Something as following:

```
if (hostOk || userOk || dbOk || portOk || pwdOk) {
    // Read connection params from source-profile.
    if (! (hostOk && userOk && dbOk ) ) {
       // Missing required connection params, return error.
    }
    // return error if mysql.host == "" etc
} else {
    // User did not specify connection params in source-profile at all, fall back to reading from environment variables.
   mysql.host = os.Getenv("MYSQLHOST")
   ....
   // return error if mysql.host == "" etc, error should mention environment variables and not params from source-profile
}
```

Reply:
Similarly for NewSourceProfileConnectionPostgreSQL.